api/v3/Contribution.php

```diff
@@ -616,84 +616,76 @@ function _civicrm_api3_contribution_completetransaction_spec(&$params) {
     'trxn_id',
-    'total_amount',
     'campaign_id',
     'fee_amount',
     'financial_type_id',
     'contribution_status_id',
     'payment_processor_id',
-  ];
-  $input = array_intersect_key($params, array_fill_keys($passThroughParams, NULL));
+    'is_email_receipt',
+    'trxn_date',
+    'receive_date',
+    'card_type_id',
+    'pan_truncation',
+    'payment_instrument_id',
+    // Q. not contribution_page_id ?
+  ];
+  $input = array_intersect_key($params, array_fill_keys($paramsToCopy, NULL));
+  // Ensure certain keys exist with NULL values if they don't already (not sure if this is ACTUALLY necessary?)
+  $input += array_fill_keys(['card_type_id', 'pan_truncation'], NULL);
+
+  // Set amount, use templateContribution if not in params.
+  $input['total_amount'] = $params['total_amount'] ?? $templateContribution['total_amount'];
+  // Why do we need this extra 'amount' key? It's used in CRM_Contribute_BAO_Contribution::completeOrder to pass on to CRM_Contribute_BAO_ContributionRecur::updateRecurLinkedPledge
+  // but it could possibly be removed if that code was adjusted unless anything else uses 'amount'
+  $input['amount'] = $input['total_amount'];
+
+  $input['payment_processor_id'] = civicrm_api3('contributionRecur', 'getvalue', [
+    'return' => 'payment_processor_id',
+    'id' => $templateContribution['contribution_recur_id'],
+  ]);
 
-  $ids = [];
-  if (!$contribution->loadRelatedObjects(['payment_processor_id' => $input['payment_processor_id']], $ids, TRUE)) {
-    throw new API_Exception('failed to load related objects');
-  }
-  unset($contribution->id, $contribution->receive_date, $contribution->invoice_id);
-  $contribution->receive_date = $params['receive_date'];
+  $input['is_test'] = $templateContribution['is_test'];
 
-  // @todo Copied from _ipn_process_transaction - needs cleanup/refactor
-  $objects = $contribution->_relatedObjects;
-  $objects['contribution'] = &$contribution;
-  $input['component'] = $contribution->_component;
-  $input['is_test'] = $contribution->is_test;
-  $input['amount'] = empty($input['total_amount']) ? $contribution->total_amount : $input['total_amount'];
-
-  if (isset($params['is_email_receipt'])) {
-    $input['is_email_receipt'] = $params['is_email_receipt'];
-  }
-  if (!empty($params['trxn_date'])) {
-    $input['trxn_date'] = $params['trxn_date'];
-  }
-  if (!empty($params['receive_date'])) {
-    $input['receive_date'] = $params['receive_date'];
-  }
-  if (empty($contribution->contribution_page_id)) {
+  if (empty($templateContribution['contribution_page_id'])) {
+    // Q. why do we need statics here?
     static $domainFromName;
     static $domainFromEmail;
     if (empty($domainFromEmail) && (empty($params['receipt_from_name']) || empty($params['receipt_from_email']))) {
       [$domainFromName, $domainFromEmail] = CRM_Core_BAO_Domain::getNameAndEmail(TRUE);
     }
-    $input['receipt_from_name'] = CRM_Utils_Array::value('receipt_from_name', $params, $domainFromName);
```

Comment:
Hmm - this might be a challenger for separating the `repeatTransaction` call from the `Payment.create` call - in that `Payment.create` doesn't take these params as yet AFAIK